In [9]:
# main data pull
import pandas as pd
import numpy as np

maincsv = "Resources/purchase_data.csv"
alldata = pd.read_csv(maincsv)
alldata.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

# Player Count

In [22]:
players = alldata["SN"].unique()
playercount = len(players)
PlayerTable = pd.DataFrame([{"Player Count": playercount}])
PlayerTable = PlayerTable.set_index("Player Count")
PlayerTable

Empty DataFrame
Columns: []
Index: [576]

# Purchasing Analysis (Total)

In [21]:
# Unique Items
items = alldata["Item ID"].unique()
itemcount = len(items)

# Total Purchases
purchases = len(alldata["Purchase ID"])

# Total Revenue
totalrev = alldata["Price"].sum()
totalrev = float(totalrev)

# Average Purchase Price
avgprice = totalrev / purchases
avgprice = float(avgprice)

# Formatting/Data Framing
totalrev = '${:2f}'.format(totalrev)
avgprice ='${:2f}'.format(avgprice)

Summary = pd.DataFrame([{"Number of Unique Items": itemcount, 
                                             "Average Purchase Price": avgprice, 
                                            "Number of Purchases": purchases, 
                                             "Total Revenue": totalrev}])

Summary = Summary[["Number of Unique Items", 
                                   "Number of Purchases", 
                                   "Average Purchase Price", 
                                   "Total Revenue"]]

Summary = Summary.set_index("Number of Unique Items")
Summary

Number of Purchases Average Purchase Price  \
Number of Unique Items                                               
183                                     780              $3.050987   

                       Total Revenue  
Number of Unique Items                
183                     $2379.770000

# Gender Demographics

In [20]:
# Data Pulls
males = alldata.loc[alldata["Gender"] == 'Male']
malecount = len(males["SN"].unique())

females = alldata.loc[alldata["Gender"] == 'Female']
femalecount = len(females["SN"].unique())

others = alldata.loc[alldata["Gender"] == 'Other / Non-Disclosed']
othercount = len(others["SN"].unique())

malePCT = malecount / playercount * 100
femalePCT = femalecount / playercount * 100
othersPCT = othercount / playercount * 100

GenderDemo = pd.DataFrame([{"Gender": "Male", 
                                                 "Total Count": malecount, 
                                                 "Percentage of Players": malePCT}, 
                                                {"Gender": "Female", 
                                                 "Total Count": femalecount, 
                                                 "Percentage of Players": femalePCT}, 
                                                {"Gender": "Other / Non-Disclosed", 
                                                 "Total Count": othercount, 
                                                 "Percentage of Players": othersPCT}])

GenderDemo = GenderDemo [["Gender", "Total Count", "Percentage of Players"]]

GenderDemo["Percentage of Players"] = GenderDemo["Percentage of Players"].map("{:.2f}%".format)
GenderDemo = GenderDemo.set_index("Gender")
GenderDemo

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

# Puchase Analysis (Gender)

In [18]:
malepurchases = len(males["Purchase ID"])
malevalue = males["Price"].sum()
malevalue = float(malevalue)

femalepurchases = len(females["Purchase ID"])
femalevalue = females["Price"].sum()
femalevalue = float(femalevalue)

otherpurchases = len(others["Purchase ID"])
othervalue = others["Price"].sum()
othervalue = float(othervalue)

maleavgprice = malevalue / malepurchases
femaleavgprice = femalevalue / femalepurchases
otheravgprice = othervalue / otherpurchases

maleavgpurchase = malevalue / malecount
femaleavgpurchase = femalevalue / femalecount
otheravgpurchase = othervalue / othercount

GenderAnalysis = pd.DataFrame([{"Gender": "Male", 
                                        "Purchase Count": malepurchases, 
                                        "Avg Purchase Price": maleavgprice,
                                       "Total Purchase Value": malevalue, 
                                        "Avg Total Purchase per Person": maleavgpurchase}, 
                                       {"Gender": "Female", "Purchase Count": femalepurchases, 
                                        "Avg Purchase Price": femaleavgprice,
                                       "Total Purchase Value": femalevalue, 
                                        "Avg Total Purchase per Person": femaleavgpurchase}, 
                                       {"Gender": "Other / Non-Disclosed", 
                                        "Purchase Count": otherpurchases, 
                                        "Avg Purchase Price": otheravgprice,
                                       "Total Purchase Value": othervalue, 
                                        "Avg Total Purchase per Person": otheravgpurchase
                                       }])

GenderAnalysis = GenderAnalysis[["Gender", "Purchase Count", "Avg Purchase Price",  "Total Purchase Value", "Avg Total Purchase per Person"]]

GenderAnalysis["Avg Purchase Price"] = GenderAnalysis["Avg Purchase Price"].map("${:.2f}".format)
GenderAnalysis["Total Purchase Value"] = GenderAnalysis["Total Purchase Value"].map("${:.2f}".format)
GenderAnalysis["Avg Total Purchase per Person"] = GenderAnalysis["Avg Total Purchase per Person"].map("${:.2f}".format)

GenderAnalysis = GenderAnalysis.set_index("Gender")
GenderAnalysis

Purchase Count Avg Purchase Price Total Purchase Value  \
Gender                                                                          
Male                              652              $3.02             $1967.64   
Female                            113              $3.20              $361.94   
Other / Non-Disclosed              15              $3.35               $50.19   

                      Avg Total Purchase per Person  
Gender                                               
Male                                          $4.07  
Female                                        $4.47  
Other / Non-Disclosed                         $4.56

# Age Demographics 

In [24]:
# Personal Note - Capitalize things. Makes reading your much easier...

ageBins = [0, 9, 14, 19, 24, 29, 34, 39, 200]
ageBinNames = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

ageDF = alldata.loc[:, ["Age", "SN", "Price"]]

ageDF["Age"] = pd.cut(ageDF["Age"], ageBins, labels=ageBinNames)

agePurchasesList = []
ageCountList = []
ageValue = []

for i in range(len(ageBinNames)):
    agePurchasesList.append(ageDF.loc[ageDF["Age"] == ageBinNames[i], :])

    ageCountList.append(len(agePurchasesList[i]["SN"].unique()))

    ageValue.append(float(agePurchasesList[i]["Price"].sum()))

    agePurchasesList[i] = len(agePurchasesList[i])

AgeDemoTable = pd.DataFrame({"Age": ageBinNames,"Players": ageCountList})
AgeDemoTable = AgeDemoTable.set_index("Players")
AgeDemoTable

# Struggled trying to figure out percentage of ages with usage of bins...

Age
Players       
17         <10
22       10-14
107      15-19
258      20-24
77       25-29
52       30-34
31       35-39
12         40+

In [27]:
#Person note - start capitalizing things. It helps make things easier to read ya dork#

ageAvgPurchasePrice = []
ageAvgPurchase = []

for i in range(len(ageBinNames)):
    ageAvgPurchasePrice.append(ageValue[i] / agePurchasesList[i])
    ageAvgPurchase.append(ageValue[i] / ageCountList[i])
    
agePurchaseAnalysis = pd.DataFrame({"Age": ageBinNames, 
                                     "Purchase Count": agePurchasesList, 
                                    "Avg Purchase Price": ageAvgPurchasePrice, 
                                     "Total Purchase Value": ageValue, 
                                    "Avg Total Purchase By Age Group": ageAvgPurchase
                                    })

agePurchaseAnalysis = agePurchaseAnalysis[["Age", 
                                           "Purchase Count", 
                                           "Avg Purchase Price", 
                                           "Total Purchase Value", 
                                           "Avg Total Purchase By Age Group"
                                          ]]

agePurchaseAnalysis["Avg Purchase Price"] = agePurchaseAnalysis["Avg Purchase Price"].map("${:.2f}".format)
agePurchaseAnalysis["Total Purchase Value"] = agePurchaseAnalysis["Total Purchase Value"].map("${:.2f}".format)
agePurchaseAnalysis["Avg Total Purchase By Age Group"] = agePurchaseAnalysis["Avg Total Purchase By Age Group"].map("${:.2f}".format)
agePurchaseAnalysis = agePurchaseAnalysis.set_index("Age")
agePurchaseAnalysis

Purchase Count Avg Purchase Price Total Purchase Value  \
Age                                                             
<10                23              $3.35               $77.13   
10-14              28              $2.96               $82.78   
15-19             136              $3.04              $412.89   
20-24             365              $3.05             $1114.06   
25-29             101              $2.90              $293.00   
30-34              73              $2.93              $214.00   
35-39              41              $3.60              $147.67   
40+                13              $2.94               $38.24   

      Avg Total Purchase By Age Group  
Age                                    
<10                             $4.54  
10-14                           $3.76  
15-19                           $3.86  
20-24                           $4.32  
25-29                           $3.81  
30-34                           $4.12  
35-39                           $4.76  
40+                             $3.19

# Top Spenders

In [29]:
playerGroupedPurchases = alldata.groupby(["SN"])
playerGroupedPurchasesDF = playerGroupedPurchases.sum()

mostSpent = playerGroupedPurchasesDF.sort_values("Price", ascending = False)
mostSpent = mostSpent.reset_index()

topSpendersList = mostSpent.iloc[0:5, 0]
topSpendersTotalSpentList = mostSpent.iloc[0:5, 4]

playerPurchaseCountList = alldata["SN"].value_counts()
playerPurchaseCountDF = pd.DataFrame(playerPurchaseCountList)

topSpendersPurchaseCount = []

for i in topSpendersList:
    playerPurchaseCount = playerPurchaseCountDF.loc[i, "SN"]
    topSpendersPurchaseCount.append(playerPurchaseCount)

topSpendersAvgPurchaseList = []

for i in range(len(topSpendersList)):    
    topSpendersAvgPurchaseList.append(topSpendersTotalSpentList[i] / topSpendersPurchaseCount[i])
    
topSpendersAnalysis = pd.DataFrame({"SN": topSpendersList, 
                                    "Purchases": topSpendersPurchaseCount, 
                                    "Total Purchase Value": topSpendersTotalSpentList, 
                                    "Avg Purchase Price": topSpendersAvgPurchaseList
                                   })

topSpendersAnalysis = topSpendersAnalysis[["SN", 
                                           "Purchases", 
                                           "Avg Purchase Price", 
                                           "Total Purchase Value"
                                          ]]

topSpendersAnalysis["Avg Purchase Price"] = topSpendersAnalysis["Avg Purchase Price"].map("${:.2f}".format)
topSpendersAnalysis["Total Purchase Value"] = topSpendersAnalysis["Total Purchase Value"].map("${:.2f}".format)

indexedTopSpendersAnalysis = topSpendersAnalysis.set_index("SN")

indexedTopSpendersAnalysis

Purchases Avg Purchase Price Total Purchase Value
SN                                                            
Lisosia93            5              $3.79               $18.96
Idastidru52          4              $3.86               $15.45
Chamjask73           3              $4.61               $13.83
Iral74               4              $3.40               $13.62
Iskadarya95          3              $4.37               $13.10

# Most Popular Items

In [32]:
itemData = alldata.loc[: , ["Item ID", "Item Name", "Price"]]

groupedItemData = itemData.groupby(["Item ID", "Item Name"])
groupedItemCount = groupedItemData.count()

sortedItemCount = groupedItemCount.sort_values(["Price"], ascending = False)

sortedItemCount = sortedItemCount.reset_index()

popularItemIDList = sortedItemCount.iloc[0:5, 0]
popularItemNameList = sortedItemCount.iloc[0:5, 1]
popularItemPurchaseCount = sortedItemCount.iloc[0:5, 2]

popularItemTotalSales = []
popularItemPrices = []

groupedItemSum = groupedItemData.sum()

for i in popularItemIDList:
    itemTotalSales = float(groupedItemSum.loc[i, "Price"])
    popularItemTotalSales.append(itemTotalSales)
    
    itemRows = alldata.loc[alldata["Item ID"] == i, :]
    itemPrice = itemRows.iloc[0, 6]
    popularItemPrices.append(itemPrice)
    
popularItemAnalysisTable = pd.DataFrame({"Item ID": popularItemIDList, 
                                         "Item Name": popularItemNameList, 
                                         "Purchases": popularItemPurchaseCount, 
                                         "Item Price": popularItemPrices, 
                                         "Total Purchase Value": popularItemTotalSales
                                         })

popularItemAnalysisTable = popularItemAnalysisTable[["Item ID", 
                                                     "Item Name", 
                                                     "Purchases", 
                                                     "Item Price", 
                                                     "Total Purchase Value"
                                                    ]]

popularItemAnalysisTable["Item Price"] = popularItemAnalysisTable["Item Price"].map("${:.2f}".format)
popularItemAnalysisTable["Total Purchase Value"] = popularItemAnalysisTable["Total Purchase Value"].map("${:.2f}".format)

indexedPopularItemAnalysisTable = popularItemAnalysisTable.set_index(["Item ID", "Item Name"])

indexedPopularItemAnalysisTable

Purchases Item Price  \
Item ID Item Name                                                            
178     Oathbreaker, Last Hope of the Breaking Storm         12      $4.23   
145     Fiery Glass Crusader                                  9      $4.58   
108     Extraction, Quickblade Of Trembling Hands             9      $3.53   
82      Nirvana                                               9      $4.90   
19      Pursuit, Cudgel of Necromancy                         8      $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

# Top Items

In [35]:
sortedItemSum = groupedItemSum.sort_values(["Price"], ascending = False)
sortedItemSum = sortedItemSum.reset_index()

topItemIDList = sortedItemSum.iloc[0:5, 0]
topItemNameList = sortedItemSum.iloc[0:5, 1]
topItemTotalSales = sortedItemSum.iloc[0:5, 2]

topItemPurchaseCount = []
topItemPrices = []

for i in topItemIDList:
    itemPurchaseCount = int(groupedItemCount.loc[i, "Price"])
    topItemPurchaseCount.append(itemPurchaseCount)

    itemRows = alldata.loc[alldata["Item ID"] == i, :]
    itemPrice = itemRows.iloc[0, 6]
    topItemPrices.append(itemPrice)    

topItemAnalysisTable = pd.DataFrame({"Item ID": topItemIDList, 
                                     "Item Name": topItemNameList, 
                                     "Purchases": topItemPurchaseCount, 
                                     "Item Price": topItemPrices, 
                                     "Total Purchase Value": topItemTotalSales
                                    })

topItemAnalysisTable = topItemAnalysisTable[["Item ID", 
                                             "Item Name", 
                                             "Purchases", 
                                             "Item Price", 
                                             "Total Purchase Value"
                                            ]]

topItemAnalysisTable["Item Price"] = topItemAnalysisTable["Item Price"].map("${:.2f}".format)
topItemAnalysisTable["Total Purchase Value"] = topItemAnalysisTable["Total Purchase Value"].map("${:.2f}".format)

indexedTopItemAnalysisTable = topItemAnalysisTable.set_index(["Item ID", "Item Name"])

indexedTopItemAnalysisTable

Purchases Item Price  \
Item ID Item Name                                                            
178     Oathbreaker, Last Hope of the Breaking Storm         12      $4.23   
82      Nirvana                                               9      $4.90   
145     Fiery Glass Crusader                                  9      $4.58   
92      Final Critic                                          8      $4.88   
103     Singed Scalpel                                        8      $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

# Heroes of Pymoil Analysis Observations

* There have been 780 total purchases made by 576 players. However, the data shows that this is showing the players who have made at least one purchase and does not encompass ALL players. Only though who have purchased items.

* 45% of players were between the ages of 20-24. In addition, this group also provided the most revenue for purchases with 365 purchases (approximately 47%) of all purchases. 

* 84% of the players were male. In addition, this gender group provided the most revenue for purchases with 652 purchases (approximately 83%) of all purchases.

* Therefore, it is highly likely the main audience Heroes of Pymoil should target is males between the ages of 20-24. This seems to be their main demographic. 
